GpsDataAnalyzer/calculator/alignment_algorithms.py

```diff
@@ -0,0 +1,293 @@
+    else:
+        print("Optimal offset: set 1 is %s seconds from set 2" % optimal_offset)
+        return (later_start_time + timedelta(seconds=optimal_offset), later_start_time)
+
+def find_lineup_naive(set1, set2):
+    """
+    Initial naive implementation to find lineup between two data sets.
+
+    Finds the point in the secondary data set closest to the start of the
+    primary data set, and then uses those starting indices to find the offset
+    with the lowest mean distance in the range of points around it index-wise
+    in the data set
+
+    Args:
+        set1: GpsDataSet object
+        set2: GpsDataSet object
+
+    Returns:
+        The indexes of the starting points in each data set
+    """
+    sets = [set1, set2]
+    starting_indexes = [0,0]
+    set1_start_time = set1.gps_data_list[0].time
+    set2_start_time = set2.gps_data_list[0].time
+    if set1_start_time > set2_start_time:
+        later_start_time = set1_start_time
+        primary_set_index = 0
+        secondary_set_index = 1
+    else:
+        later_start_time = set2_start_time
+        primary_set_index = 1
+        secondary_set_index = 0
+
+    smallest_time_difference = None
+    for i,data_point in enumerate(sets[secondary_set_index].gps_data_list):
+        time_difference = abs((data_point.time - later_start_time).total_seconds())
+        if smallest_time_difference is None or time_difference < smallest_time_difference:
+            smallest_time_difference = time_difference
+            starting_indexes[secondary_set_index] = i
+    best_index = 0
+    best_mean_distance = None
+    for offset in range(-100,100):
+        distances = []
+        moving_index = starting_indexes[secondary_set_index] + offset
+        if moving_index >= 0:
+            primary_points = sets[primary_set_index].gps_data_list
+            secondary_points = sets[secondary_set_index].gps_data_list[moving_index:]
+            for primary_point,secondary_point in zip(primary_points, secondary_points):
+                location1 = (primary_point.latitude, primary_point.longitude)
+                location2 = (secondary_point.latitude, secondary_point.longitude)
+                distances.append(utils.calculate_distance(location1, location2))
+        if distances and (best_mean_distance == None or np.mean(distances) < best_mean_distance):
+            best_mean_distance = np.mean(distances)
+            best_index = moving_index
+    print("optimal mean distance: " + str(best_mean_distance))
+    starting_indexes[secondary_set_index] = best_index
+    return starting_indexes
+
+def create_time_to_points_mapping(dataset, offset=0):
+    """
+    Map common timestamp to points from both sets.
+
+    Args:
+        dataset: GpsDataSet
+        offset: int, seconds of offset to apply to timestamps
+
+    Returns:
+        Dictionary that maps datetime timestamps to the points that fall at that
```

Comment:
maybe another way to explain this is: bucketize/group dataset by seconds timestamp?